---
title: "towhee.introgress"
author: "Devon DeRaad"
date: "2/11/2022"
output: html_document
---

```{r, message=FALSE}
library(ggplot2)
library(introgress)
library(vcfR)
library(SNPfiltR)
library(adegenet)
```

```{r}
#read in vcfR
vcfR<-read.vcfR("/Users/devder/Desktop/mex.towhees/towhee.75.mac2.vcf")
#read in sample info csv
sample.info<-read.csv("~/Desktop/mex.towhees/sample.locs.csv")

#make sure sampling file matches vcf
sample.info$ID %in% colnames(vcfR@gt)[-1]
colnames(vcfR@gt)[-1] %in% sample.info$ID

#if needed, subset sampling file to include only samples in vcf
#sample.info<-sample.info[sample.info$id %in% colnames(vcfR@gt)[-1],]

#reorder sampling file to match order of samples in vcf
sample.info<-sample.info[match(colnames(vcfR@gt)[-1], sample.info$ID),]
sample.info$ID == colnames(vcfR@gt)[-1]

#mac5<-min_mac(vcfR, min.mac = 5)
#write.vcf(mac5, file="~/Downloads/mac5.vcf.gz")
```

```{r}
#convert to genlight
gen<-vcfR2genlight(vcfR)
#perform PCA
pca<-glPca(gen, nf=4)
var_frac <- pca$eig/sum(pca$eig)

#pull pca scores out of df
pca.scores<-as.data.frame(pca$scores)
pca.scores$locat<-sample.info$Distance
pca.scores$site<-sample.info$Location

#ggplot color by species
ggplot(pca.scores, aes(x=PC1, y=PC2, color=site)) +
  geom_point(cex = 5, alpha=.5)+
  theme_classic()
```

```{r}
#create SNP matrix
mat<-extract.gt(vcfR)
mat[1:5,1:5]
dim(mat)
colnames(mat)

#convert matrix to numeric
conv.mat<-mat
conv.mat[conv.mat == "0/0"]<-0
conv.mat[conv.mat == "0/1"]<-1
conv.mat[conv.mat == "1/1"]<-2
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}

#calc AF for the samples you will use to call fixed differences
ocai.af<-(rowSums(conv.mat[,c(7:10)], na.rm=T)/(rowSums(is.na(conv.mat[,c(7:10)]) == FALSE)))/2
maculatus.af<-(rowSums(conv.mat[,c(1,2,19,20)], na.rm=T)/(rowSums(is.na(conv.mat[,c(1,2,19,20)]) == FALSE)))/2

#find fixed SNPs
diff<-abs(ocai.af - maculatus.af)
#how many SNPs are fixed
table(is.na(diff) == FALSE & diff == 1) #20
#how many have > .8 allele frequency diff
table(is.na(diff) == FALSE & diff > .8) #37
```

```{r}
#subsample original matrix to only fixed diff SNPs
gen.mat<-mat[is.na(diff) == FALSE & diff > .8,]
dim(gen.mat)

#subsample matrix converted for AF calcs to only fixed SNPS
conv.mat<-conv.mat[is.na(diff) == FALSE & diff > .8,]
dim(conv.mat)
```

```{r}
#write a logical test to convert alleles so that a single number represents one parental ancestry
for (i in 1:nrow(gen.mat)){
  #if 1 is the ocai allele (ie <= .4 frequency in the 4 maculatus samples used for identifying informative SNPs)
if((sum(conv.mat[i,c(1,2,19,20)], na.rm=T)/(sum(is.na(conv.mat[i,c(1,2,19,20)]) == FALSE)))/2 < .2){
  #swap all '0/0' cells with '2/2'
    gen.mat[i,][gen.mat[i,] == "0/0"]<-"2/2"
    #swap all '1/1' cells with '0/0'
    gen.mat[i,][gen.mat[i,] == "1/1"]<-"0/0"
    #finally convert all '2/2' cells (originally 0/0) into '1/1'
    gen.mat[i,][gen.mat[i,] == "2/2"]<-"1/1"
    #no need to touch hets
  }
}

#convert R class NAs to the string "NA/NA"
gen.mat[is.na(gen.mat) == TRUE]<-"NA/NA"

#make locus info df
locus.info<-data.frame(locus=rownames(gen.mat),
                       type=rep("C", times=nrow(gen.mat)),
                       lg=vcfR@fix[,1][is.na(diff) == FALSE & diff > .8],
                       marker.pos=vcfR@fix[,2][is.na(diff) == FALSE & diff > .8])
#make linkage group numeric
locus.info$lg<-substr(locus.info$lg,5,15)
rs<-unlist(strsplit(locus.info$lg, split = "_"))
rs<-rs[c(TRUE, FALSE)]
locus.info$lg<-rs
locus.info$lg<-as.numeric(as.character(locus.info$lg))
```

```{r}
#we now have a gt matrix in proper format for introgress
#convert genotype data into a matrix of allele counts
count.matrix<-prepare.data(admix.gen=gen.mat, loci.data=locus.info,
                           parental1="1",parental2="0", pop.id=F,
                           ind.id=F, fixed=T)

#estimate hybrid index values
hi.index.sim<-est.h(introgress.data=count.matrix,loci.data=locus.info,
                    fixed=T, p1.allele="1", p2.allele="0")

#make plot
mk.image(introgress.data=count.matrix, loci.data = locus.info,
         hi.index=hi.index.sim, ylab.image="Individuals",
         marker.order=order(locus.info$lg), xlab.h="population 2 ancestry", pdf=F,
         col.image=c("#B2182B","black","#4D4D4D"))
```

### Out of our 20 fixed differences between cline endpoints, 12 occur on a single UCE locus, 'uce-5216_MVZ122216'. Blasting the reference contig for this UCE locus against the Zebra Finch reference genome gives a top hit on the mitochondria. This must be a misassembled UCE or nuMT? We will need to blast all fixed difference UCEs and make sure they aren't mitochondrial.
```{r}
#remove mitochondrial locus from gen.mat and locus.info
gen.mat<-gen.mat[15:37,]
locus.info<-locus.info[15:37,]

#we now have a gt matrix in proper format for introgress
#convert genotype data into a matrix of allele counts
count.matrix<-prepare.data(admix.gen=gen.mat, loci.data=locus.info,
                           parental1="1",parental2="0", pop.id=F,
                           ind.id=F, fixed=T)

#estimate hybrid index values
hi.index.sim<-est.h(introgress.data=count.matrix,loci.data=locus.info,
                    fixed=T, p1.allele="1", p2.allele="0")

#make plot
mk.image(introgress.data=count.matrix, loci.data = locus.info,
         hi.index=hi.index.sim, ylab.image="Individuals",
         marker.order=order(locus.info$lg), xlab.h="population 2 ancestry", pdf=F,
         col.image=c(rgb(1,0,0,alpha=.5),rgb(0,0,0,alpha=.8),rgb(0,0,1,alpha=.5)))

#save df
g<-as.data.frame(count.matrix[["Admix.gen"]])
#write df out
write.csv(g, "~/Desktop/mex.towhees/diagnostic.snps.csv", quote=F, row.names = F)
```


```{r}
#calculate mean heterozygosity across these diagnostic markers for each sample
#using their function
het<-calc.intersp.het(introgress.data=count.matrix)
#dev.off()
#make triangle plot
introgress::triangle.plot(hi.index=hi.index.sim, int.het=het, pdf = F)
```

```{r}
#make better triangle plot with color-coding according to sampling locality
#plot triangle
plot(x=hi.index.sim$h, y=het, bg=c(rep(rgb(1,0,0,alpha=.5), times=2), rep(rgb(1,1,0,alpha=.5), times=4),
                                   rep(rgb(0,1,0,alpha=.5), times=4), rep(rgb(0,0,1,alpha=.5), times=4),
                                   rep(rgb(1,1,1,alpha=.5), times=4), rep(rgb(1,0,1,alpha=.5), times=2),
                                   rep(rgb(0,1,1,alpha=.5), times=4), rep(rgb(1,.5,0,alpha=.5), times=4),
                                   rep(rgb(1,0,.5,alpha=.5), times=4)),
     pch=21, cex=1.5,
     xlab="Hybrid Index", ylab="Interspecific heterozygosity",
     ylim=c(0,1))
segments(x0 =0, y0 =0, x1 =.5, y1 =1)
segments(x0 =1, y0 =0, x1 =.5, y1 =1)

#make data frame to hold this info
d<-data.frame(hi.index=hi.index.sim$h, heterozygosity=het,
              sample=colnames(count.matrix[["Admix.gen"]]),
              loc=sample.info$sample.loc)

#write this df out
write.csv(d, "~/Desktop/mex.towhees/triangle.plot.df.csv", quote=F, row.names = F)

```

```{r}
#remove mitochondrial scaffold 'uce-5216_MVZ122216' from uce vcf
#vcfR@fix #check where it is to subset
vcfR<-vcfR[c(1:40,65:5028),]
#vcfR@fix #make sure the subset worked

#write out new vcf
#vcfR::write.vcf(vcfR, file = "~/Desktop/mex.towhees/towhee.75.mac2.nomito.vcf.gz")
#perform linkage filtering to get a reduced vcf with only unlinked (one per UCE) SNPs
vcfR.thin<-distance_thin(vcfR, min.distance = 10000)
vcfR.thin #retains 2,668 SNPs
#vcfR::write.vcf(vcfR.thin, file = "~/Desktop/mex.towhees/towhee.75.mac2.unlinked.nomito.vcf.gz")
```

