---
title: "towhee.popgen"
author: "Devon DeRaad"
date: "2/12/2022"
output: html_document
---

### Read in data
```{r}
#install.packages("elevatr")
#remotes::install_github("wmgeolab/rgeoboundaries")
#load packages
library(vcfR)
library(ggplot2)
library(adegenet)
library(SNPfiltR)
library(StAMPP)
library(viridis)
library(rgeoboundaries)
library(elevatr)
library(raster)
library(sf)

#check out the details on the filtered vcf
vcfR <- read.vcfR("~/Desktop/mex.towhees/towhee.75.mac2.nomito.vcf")
vcfR

#read in sample info csv
sample.info<-read.csv("~/Desktop/mex.towhees/sample.locs.csv")

#make sure sampling file matches vcf
sample.info$ID %in% colnames(vcfR@gt)[-1]
colnames(vcfR@gt)[-1] %in% sample.info$ID

#if needed, subset sampling file to include only samples in vcf
#sample.info<-sample.info[sample.info$id %in% colnames(vcfR@gt)[-1],]

#reorder sampling file to match order of samples in vcf
sample.info<-sample.info[match(colnames(vcfR@gt)[-1], sample.info$ID),]
sample.info$ID == colnames(vcfR@gt)[-1]
sample.info$sample.loc<-as.character(sample.info$sample.loc)
```

### Make color coded sampling map (pretty one for Fig. 1 and empty one for SNAPP figure)
```{r}
#make dataframe with only the 8 unique sampling locs
unq<-sample.info[!duplicated(sample.info$lat),]

#download elevation data for Mexico
mex_bound <- geoboundaries("Mexico")
elevation_data <- get_elev_raster(locations = mex_bound, z = 6, clip = "locations")
elevation_data <- as.data.frame(elevation_data, xy = TRUE)
colnames(elevation_data)[3] <- "elevation"
# remove rows of data frame with one or more NA's,using complete.cases
elevation_data <- elevation_data[complete.cases(elevation_data), ]

#make subset plotting area corresponding to our sampling
#subset elevation data based on latitude and longitude limits
elevation_data<-elevation_data[elevation_data$x >= -105.5 & elevation_data$x <= -96 &
                              elevation_data$y >= 16 & elevation_data$y <= 23,]

#subset mexico map outline based on limits
mex_bound <- st_crop(mex_bound, xmin = -105.5, xmax = -96,
                                    ymin = 16, ymax = 23)

#plot the map with ggplot
ggplot() +
  geom_raster(data = elevation_data, aes(x = x, y = y, fill = elevation)) +
  geom_sf(data = mex_bound, color = "black", fill = NA, cex=.3) +
  geom_point(data = unq, aes(x = long, y = lat, color=sample.loc), size=10, alpha =.8, show.legend=FALSE) +
  scale_color_brewer(palette = "RdGy")+
  geom_text(data = unq, aes(x = long, y = lat, label = sample.loc), size = 6, color="white")+
  #scale_fill_gradient(low = "white", high = "black")+
  scale_fill_gradientn(colours = terrain.colors(50))+
  labs(x = "Longitude", y = "Latitude", fill = "Elevation\n(meters)")+
  theme_classic()+
  theme(legend.position = c(0.01, 0.01), legend.justification = c(0.01, 0.01),
        legend.background = element_blank())


#plot the map with ggplot
ggplot() +
  geom_raster(data = elevation_data, aes(x = x, y = y, fill = elevation)) +
  geom_sf(data = mex_bound, color = "black", fill = NA, cex=.3) +
  geom_point(data = unq, aes(x = long, y = lat, color=sample.loc), size=10, alpha =.8, show.legend=FALSE) +
  scale_color_brewer(palette = "PuOr")+
  geom_text(data = unq, aes(x = long, y = lat, label = sample.loc), size = 6, color="white")+
  scale_fill_gradient(low = "white", high = "black")+
  #scale_fill_gradientn(colours = terrain.colors(50))+
  labs(x = "Longitude", y = "Latitude", fill = "Elevation\n(meters)")+
  theme_classic()+
  theme(legend.position = c(0.01, 0.01), legend.justification = c(0.01, 0.01),
        legend.background = element_blank())

ggplot() +
  geom_raster(data = elevation_data, aes(x = x, y = y, fill = elevation)) +
  geom_sf(data = mex_bound, color = "black", fill = NA, cex=.3) +
  geom_point(data = unq, aes(x = long, y = lat, color=sample.loc), size=10, alpha =.8, show.legend=FALSE) +
  scale_color_brewer(palette = "BrBG")+
  geom_text(data = unq, aes(x = long, y = lat, label = sample.loc), size = 6, color="white")+
  scale_fill_gradient(low = "white", high = "black")+
  #scale_fill_gradientn(colours = terrain.colors(50))+
  labs(x = "Longitude", y = "Latitude", fill = "Elevation\n(meters)")+
  theme_classic()+
  theme(legend.position = c(0.01, 0.01), legend.justification = c(0.01, 0.01),
        legend.background = element_blank())

```

```{r}
#make empty map to overlay snapp trees on
pac<-map_data("world")
#
ggplot()+
  geom_polygon(data = pac, aes(x=long, y = lat, group = group), fill="seashell", col="black", cex=.3)+
  coord_sf(xlim = c(-105.5, -96), ylim = c(16, 23)) + 
  geom_point(data = unq, aes(x = long, y = lat, color=sample.loc), size=10, alpha =.8, show.legend=TRUE) +
  scale_color_brewer(palette = "RdGy")+
  theme_classic()+
  geom_text(data = unq, aes(x = long, y = lat, label = sample.loc), size = 6, color="white")+
  theme(legend.position = "none")+
  xlab("Longitude")+
  ylab("Latitude")

ggplot()+
  geom_polygon(data = pac, aes(x=long, y = lat, group = group), fill="oldlace", col="black", cex=.3)+
  coord_sf(xlim = c(-105.5, -96), ylim = c(16, 23)) + 
  geom_point(data = unq, aes(x = long, y = lat, color=sample.loc), size=10, alpha =.8, show.legend=TRUE) +
  scale_color_brewer(palette = "RdGy")+
  theme_classic()+
  geom_text(data = unq, aes(x = long, y = lat, label = sample.loc), size = 6, color="white")+
  theme(legend.position = "none")+
  xlab("Longitude")+
  ylab("Latitude")

```


### Make PCA for both morphology and genome, each colored by sampling locality matching map
```{r}
#morph
morph.pca <- prcomp(sample.info[,c(8:13)], center = TRUE,scale. = TRUE)
sample.info$pc1morph<-morph.pca$x[,1]
sample.info$pc2morph<-morph.pca$x[,2]
ggplot(sample.info, aes(x=pc1morph, y=pc2morph, color=sample.loc)) +
  geom_point(cex=4, alpha=.8)+
  theme_classic()+
  scale_color_brewer(palette = "RdGy")

#plumage
morph.pca <- prcomp(sample.info[,c(16:21)], center = TRUE,scale. = TRUE)
sample.info$pc1plum<-morph.pca$x[,1]
sample.info$pc2plum<-morph.pca$x[,2]
ggplot(sample.info, aes(x=pc1plum, y=pc2plum, color=sample.loc)) +
  geom_point(cex=4, alpha=.8)+
  theme_classic()+
  scale_color_brewer(palette = "RdGy")

#plot genetic pca
popm<-data.frame(id=sample.info$ID, pop=sample.info$sample.loc)
gen.pc<-assess_missing_data_pca(vcfR, popmap = popm, clustering = FALSE)

#add genetic PC1 & PC2 to sample info dataframe
sample.info$genpc1<-gen.pc$PC1
sample.info$genpc2<-gen.pc$PC2

#plot genetic PCA
ggplot(sample.info, aes(x=genpc1, y=genpc2, color=sample.loc)) +
  geom_point(cex=4, alpha=.8)+
  theme_classic()+
  scale_color_brewer(palette = "RdGy")

#plot genetic PC1 against plumage PC1
ggplot(sample.info, aes(x=genpc1, y=pc1plum, color=sample.loc)) +
  geom_point(cex=4, alpha=.8)+
  theme_classic()+
  scale_color_brewer(palette = "RdGy")

#plot genetic PC1 against morph PC1
ggplot(sample.info, aes(x=genpc1, y=pc1morph, color=sample.loc)) +
  geom_point(cex=4, alpha=.8)+
  theme_classic()+
  #geom_smooth(method=lm)+
  scale_color_brewer(palette = "RdGy")

```


### add Splitstree
```{r}
#convert to genlight
gen<-vcfR2genlight(vcfR)

pop(gen)<-gen@ind.names
sample.div.mac.75 <- stamppNeisD(gen, pop = FALSE)
#export for splitstree
#stamppPhylip(distance.mat=sample.div.mac.75, file="~/Desktop/mex.towhees/towhee.75.mac.splits.txt")

#75% completeness cutoff with MAC splitstree
knitr::include_graphics(c("/Users/devder/Desktop/mex.towhees/towhee.75.mac.nomito.splitstree.png"))
#add dots color coded by locality in photoshop
```

### Make pairwise Fst heatmap
```{r}
#calc pairwise Fst
gen@pop<-as.factor(sample.info$sample.loc)
di.heat<-stamppFst(gen)
m<-di.heat$Fsts
#fill in upper triangle of matrix
m[upper.tri(m)] <- t(m)[upper.tri(m)]

#melt for plotting
heat <- reshape::melt(m)

#plot with labels
ggplot(data = heat, aes(x=X1, y=X2, fill=value)) + 
  geom_tile()+
  geom_text(data=heat,aes(label=round(value, 2)))+
  theme_minimal()+
  scale_fill_gradient2(low = "white", high = "red", space = "Lab", name="Fst") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

#identify the number of fixed differences between pops
mat<-extract.gt(vcfR)
conv.mat<-mat
conv.mat[conv.mat == "0/0"]<-0
conv.mat[conv.mat == "0/1"]<-1
conv.mat[conv.mat == "1/1"]<-2
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}

#show colnames to verify you're subsetting correctly
colnames(conv.mat) == locs$id

#make vector to fill with fixed diff values
f<-c()

#write for loop to calc number of fixed diffs between each pop
for (i in 1:nrow(heat)){
  #calc af of pop1 and pop2
  pop1.af<-(rowSums(conv.mat[,locs$species == heat$X1[i]], na.rm=T)/(rowSums(is.na(conv.mat[,locs$species == heat$X1[i]]) == FALSE)))/2
  pop2.af<-(rowSums(conv.mat[,locs$species == heat$X2[i]], na.rm=T)/(rowSums(is.na(conv.mat[,locs$species == heat$X2[i]]) == FALSE)))/2
  #store number of fixed differences
  f[i]<-sum(is.na(abs(pop1.af - pop2.af)) == FALSE & abs(pop1.af - pop2.af) == 1) #find fixed SNPs and add to vector
}

#add number of fixed diffs to df
heat$fixed<-f

#fix the assignments
heat$mixed<-heat$value
heat$mixed[c(2:7,10:14,18:21,26:28,34:35,42)]<-heat$fixed[c(2:7,10:14,18:21,26:28,34:35,42)]

#plot with labels
fst.plot<-ggplot(data = heat, aes(x=X1, y=X2, fill=value)) + 
  geom_tile()+
  scale_x_discrete(limits=levels(heat$X2)[c(1,7,2,3,4,5,6)], labels=c("1:10","12:18","26","11","24:25","19","20:23"))+ 
  scale_y_discrete(limits=levels(heat$X2)[c(1,7,2,3,4,5,6)], labels=c("1:10","12:18","26","11","24:25","19","20:23"))+
  geom_text(data=heat,aes(label=round(mixed, 2)), size=2.5)+
  theme_minimal()+
  scale_fill_gradient2(low = "white", high = "red", space = "Lab", name="Fst") +
  theme(axis.text.x = element_text(angle = 45, vjust=.25, size=12),
        axis.text.y = element_text(hjust = -.2, size=12),
        axis.title.x = element_blank(), axis.title.y = element_blank())

```


```{r}
#make figure that is 
#|_A_|_B_|
#|_C_|_D_|
#|_E_|_F_|
#|_G_|_H_|
#where ABCD = sampling map
#E = plumage PCA
#F = genomic PCA
#G = triangle plot
#H = pairwise Fst heatmap
#consider adding admixture plot in photoshop?



```



### Plot heterozygosity and Pi by sampling site
### Do we see expected pattern of increased diversity and heterozygosity in the middle of the cline?
```{r}

```

