---
title: "Untitled"
author: "Devon DeRaad"
date: "2/5/2022"
output: html_document
---

```{r}
library(vcfR)
library(SNPfiltR)

vcf<-read.vcfR("~/Desktop/mex.towhees/genotyped_mito_samples_only_PASS_snp.vcf")
#read in locality info for samples
locs<-read.csv("~/Downloads/Towhee.genetic.samples.csv", header=T)
head(vcf@fix)
vcf@gt[1:10,1:10]
#check that sampling IDs in text file match sample IDs in the vcf
locs$sample.id<-paste0(locs$Collection,locs$ID)
colnames(vcf@gt)[-1] %in% locs$sample.id
#reorder locs df to match order of vcf
locs<-locs[order(match(locs$sample.id, colnames(vcf@gt)[-1])), ]
```

```{r}
#write a function that takes a vcfR, filters heterozygote genotypes based on allele balance (allowing user to specify cutoff) and returns fitlered vcfR

abfilt <- function(vcfR,
                   min.ratio=NULL,
                   max.ratio=NULL){

  #if specified vcfR is not class 'vcfR', fail gracefully
  if (class(vcfR) != "vcfR"){
    stop("specified vcfR object must be of class 'vcfR'")
  }

  #set default parameter for minimum allele ratio for a het call (.25)
  if(is.null(min.ratio)){
    min.ratio=.25
  }
  
  #set default parameter for maximum allele ratio for a het call (.75)
    if(is.null(max.ratio)){
    max.ratio=.75
  }
  #if these ratios have been specified by the user, leave them alone
  
  #extract AD from the vcf
  ad.matrix<- vcfR::extract.gt(vcfR, element='AD')
  #extract GT from the vcf
  gt.matrix<- vcfR::extract.gt(vcfR, element='GT')

  #mask dp matrix to include only called hets from gt matrix
  ad.matrix[gt.matrix != "0/1"]<-NA

  #split allele 1 depth from allele 2 depth
  al1<-structure(as.numeric(gsub(",.*", "", ad.matrix)), dim=dim(ad.matrix))
  al2<-structure(as.numeric(gsub(".*,", "", ad.matrix)), dim=dim(ad.matrix))

  #calculate AB for each sample
  al.bal<-al1/(al1 + al2)

  #calculate logical storing whether each het genotype passes the filter
  AB<-al1/(al1 + al2) > max.ratio | al1/(al1 + al2) < min.ratio

  #calculate percent of het genotypes failing the filter
  p<-round(sum(AB, na.rm = TRUE) / sum(is.na(AB) == FALSE)*100, 2)

  #calculate overall percentage of genotypes failing the filter
  j<-round(sum(AB, na.rm = TRUE) / sum(is.na(gt.matrix) == FALSE)*100, 2)

  #print to user
  message(p,"% of het genotypes (",j,"% of all genotypes) fall outside of ",min.ratio," - ",max.ratio, " allele balance ratio and were converted to NA")

  #convert failing genotypes to NA
  vcfR@gt[,-1][AB]<-NA

  #make histogram of allele balance at all het genotypes
  graphics::hist(al.bal,
       xlim = c(0,1),
       ylab = "number of genotypes",
       xlab = "Allele balance",
       main ="allele balance distribution")
  graphics::abline(v=c(min.ratio,max.ratio),
         col="red")

  #return vcfR
  return(vcfR)

#close function
}

```

```{r}
#convert all het calls to NA
vcf<-abfilt(vcf, min.ratio = .5, max.ratio = .4)
#filter biallelic
vcf<-filter_biallelic(vcf)
#filter out singletons
vcf<-min_mac(vcf, min.mac = 2)
#assess missing data by sample
missing_by_sample(vcf)

#make popmap
pop<-locs[,c(16,8)]
colnames(pop)<-c("id","pop")
assess_missing_data_pca(vcf, popmap = pop, thresholds = c(.8,.9,1), clustering=FALSE)
```

```{r}
#filter to 100% completeness per SNP threshold
vcf.100<-missing_by_snp(vcf, cutoff = 1) #27 SNPs with no missing data

#export mitochondrial vcf
#vcfR::write.vcf(vcf.100, file="~/Desktop/mex.towhees/towhee.mito.100.vcf.gz")
```

